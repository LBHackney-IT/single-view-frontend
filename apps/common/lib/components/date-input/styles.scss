@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-date-input/date-input";

fieldset.lbh-form-group > .lbh-date-input {
  margin-top: 5px;
}
