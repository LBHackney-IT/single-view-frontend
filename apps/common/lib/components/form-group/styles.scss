@import "../../../node_modules/lbh-frontend/lbh/base";
@import "../../../node_modules/govuk-frontend/govuk/objects/form-group";
@import "../../../node_modules/lbh-frontend/lbh/objects/form-group";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-label/label";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-hint/hint";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-error-message/error-message";

.lbh-form-group {
  label,
  legend {
    sup {
      color: lbh-colour("lbh-error");
    }
  }
}

fieldset.lbh-form-group legend {
  display: contents;
  margin-bottom: 0.375rem;
}
