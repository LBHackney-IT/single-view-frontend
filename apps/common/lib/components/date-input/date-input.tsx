import React, { ComponentPropsWithoutRef, forwardRef } from "react";

import cn from "classnames";

import { NumberInput, NumberInputProps } from "../number-input";
import "./styles.scss";

export interface DateInputProps extends ComponentPropsWithoutRef<"div"> {
  id?: string;
  error?: string;
  required?: boolean;
  dayProps?: NumberInputProps;
  monthProps?: NumberInputProps;
  yearProps?: NumberInputProps;
  dayLabel?: string;
  monthLabel?: string;
  yearLabel?: string;
}

export const DateInput = forwardRef<HTMLDivElement, DateInputProps>(function DateInput(
  {
    id = "date-input",
    dayProps,
    monthProps,
    yearProps,
    dayLabel = "Day",
    monthLabel = "Month",
    yearLabel = "Year",
    error,
    required,
    className,
    ...props
  },
  ref,
) {
  return (
    <div
      ref={ref}
      className={cn("govuk-date-input", "lbh-date-input", className)}
      {...props}
    >
      <div className="govuk-date-input__item">
        <label className="govuk-label lbh-label" htmlFor={`${id}-day`}>
          {dayLabel}
        </label>
        <NumberInput
          className="govuk-date-input__input govuk-input--width-2"
          name="day"
          required={required}
          maxLength={2}
          min={1}
          max={31}
          padStart={2}
          aria-label="Day"
          {...dayProps}
        />
      </div>
      <div className="govuk-date-input__item">
        <label className="govuk-label lbh-label" htmlFor={`${id}-month`}>
          {monthLabel}
        </label>
        <NumberInput
          className="govuk-date-input__input govuk-input--width-2"
          name="month"
          required={required}
          maxLength={2}
          min={1}
          max={12}
          padStart={2}
          aria-label="Month"
          {...monthProps}
        />
      </div>
      <div className="govuk-date-input__item">
        <label className="govuk-label lbh-label" htmlFor={`${id}-year`}>
          {yearLabel}
        </label>
        <NumberInput
          className="govuk-input govuk-date-input__input govuk-input--width-4"
          name="year"
          required={required}
          maxLength={4}
          padStart={4}
          aria-label="Year"
          {...yearProps}
        />
      </div>
    </div>
  );
});
