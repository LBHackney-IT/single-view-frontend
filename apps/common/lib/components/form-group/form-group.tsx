import React, {
  Children,
  ReactElement,
  cloneElement,
  forwardRef,
  isValidElement,
  useMemo,
} from "react";

import classNames from "classnames";

import { widthOverrides } from "../../utils";
import { TextArea } from "../text-area";

import type * as Polymorphic from "@radix-ui/react-polymorphic";
import "./styles.scss";

export interface FormGroupProps {
  id: string;
  label?: string;
  name?: string;
  hint?: string;
  error?: string | false;
  required?: boolean;
  children: ReactElement;
  override?: number;
}

export type FormGroupComponent = Polymorphic.ForwardRefComponent<"div", FormGroupProps>;

export const FormGroup: FormGroupComponent = forwardRef(function FormGroup(
  {
    as: FormGroupComp = "div",
    id,
    name,
    label,
    hint,
    error,
    required,
    children,
    className,
    override,
    ...props
  },
  ref,
) {
  const formGroupClasses = classNames(
    "govuk-form-group",
    {
      "govuk-form-group--error": !!error,
    },
    "lbh-form-group",
    widthOverrides(override),
    className,
  );

  const describedBy = useMemo(() => {
    const classes: string[] = [];
    if (hint) {
      classes.push(`${id}-hint`);
    }
    if (error) {
      classes.push(`${id}-error`);
    }
    return classes.join(" ");
  }, [id, hint, error]);

  const LabelComp =
    typeof FormGroupComp === "string" && (FormGroupComp as string) === "fieldset"
      ? "legend"
      : "label";

  const formGroup = (
    <FormGroupComp ref={ref} id={id} className={formGroupClasses} {...props}>
      <LabelComp className="govuk-label lbh-label" htmlFor={`${id}-field`}>
        {label}
        {required ? <sup aria-hidden="true">*</sup> : ""}
      </LabelComp>
      {!!hint && (
        <span id={`${id}-hint`} className="govuk-hint lbh-hint">
          {hint}
        </span>
      )}
      {!!error && (
        <span id={`${id}-error`} className="govuk-error-message lbh-error-message">
          <span className="govuk-visually-hidden">Error:</span> {error}
        </span>
      )}
      {!!children &&
        Children.only(
          cloneElement(children, {
            id: `${id}-field`,
            name,
            required,
            error: !!error,
            "aria-describedby": describedBy || undefined,
            ...children.props,
          }),
        )}
    </FormGroupComp>
  );

  return isValidElement(children) && children.type === TextArea ? (
    <div className="govuk-character-count">{formGroup}</div>
  ) : (
    formGroup
  );
});
