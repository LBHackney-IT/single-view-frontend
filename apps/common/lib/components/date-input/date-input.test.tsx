import React from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";

import { FormGroup } from "../form-group";
import { DateInput } from "./date-input";

test("it passes a11y", async () => {
  const { container } = render(
    <FormGroup id="test" name="test" label="DateInput Test">
      <DateInput />
    </FormGroup>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders correctly", () => {
  const { container } = render(<DateInput />);
  expect(container).toMatchSnapshot();
});
