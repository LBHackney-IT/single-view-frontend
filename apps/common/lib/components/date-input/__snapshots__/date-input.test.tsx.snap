// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it passes a11y 1`] = `
<div>
  <div
    class="govuk-form-group lbh-form-group"
    id="test"
  >
    <label
      class="govuk-label lbh-label"
      for="test-field"
    >
      DateInput Test
      
    </label>
    <div
      class="govuk-date-input lbh-date-input"
      name="test"
    >
      <div
        class="govuk-date-input__item"
      >
        <label
          class="govuk-label lbh-label"
          for="test-field-day"
        >
          Day
        </label>
        <input
          aria-label="Day"
          aria-valuemax="31"
          aria-valuemin="1"
          class="govuk-input lbh-input govuk-date-input__input govuk-input--width-2"
          name="day"
          role="spinbutton"
          value=""
        />
      </div>
      <div
        class="govuk-date-input__item"
      >
        <label
          class="govuk-label lbh-label"
          for="test-field-month"
        >
          Month
        </label>
        <input
          aria-label="Month"
          aria-valuemax="12"
          aria-valuemin="1"
          class="govuk-input lbh-input govuk-date-input__input govuk-input--width-2"
          name="month"
          role="spinbutton"
          value=""
        />
      </div>
      <div
        class="govuk-date-input__item"
      >
        <label
          class="govuk-label lbh-label"
          for="test-field-year"
        >
          Year
        </label>
        <input
          aria-label="Year"
          class="govuk-input lbh-input govuk-input govuk-date-input__input govuk-input--width-4"
          name="year"
          role="spinbutton"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`it renders correctly 1`] = `
<div>
  <div
    class="govuk-date-input lbh-date-input"
  >
    <div
      class="govuk-date-input__item"
    >
      <label
        class="govuk-label lbh-label"
        for="date-input-day"
      >
        Day
      </label>
      <input
        aria-label="Day"
        aria-valuemax="31"
        aria-valuemin="1"
        class="govuk-input lbh-input govuk-date-input__input govuk-input--width-2"
        name="day"
        role="spinbutton"
        value=""
      />
    </div>
    <div
      class="govuk-date-input__item"
    >
      <label
        class="govuk-label lbh-label"
        for="date-input-month"
      >
        Month
      </label>
      <input
        aria-label="Month"
        aria-valuemax="12"
        aria-valuemin="1"
        class="govuk-input lbh-input govuk-date-input__input govuk-input--width-2"
        name="month"
        role="spinbutton"
        value=""
      />
    </div>
    <div
      class="govuk-date-input__item"
    >
      <label
        class="govuk-label lbh-label"
        for="date-input-year"
      >
        Year
      </label>
      <input
        aria-label="Year"
        class="govuk-input lbh-input govuk-input govuk-date-input__input govuk-input--width-4"
        name="year"
        role="spinbutton"
        value=""
      />
    </div>
  </div>
</div>
`;
